### PR TITLE
fix: nested scan (#6136)

### PR DESCRIPTION
**Pull request by black-06**
- [X] Do only one thing
- [X] Non breaking API changes
- [X] Tested

### What did this pull request do?

fix nested scan panic #6136 

### User Case Description

```go
var user User
if err := DB.Scopes(func(db *gorm.DB) *gorm.DB {
	var maxID int64
	if err := db.Raw("select max(id) from users").Scan(&maxID).Error; err != nil {
		return db
	}
	return db.Raw("select * from users where id = ?", maxID)
}).Scan(&user).Error; err != nil {
	t.Errorf("failed to find user, got err: %v", err)
}
```

### CODE REVIEW

File: statement.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ type Statement struct {
 	attrs                []interface{}
 	assigns              []interface{}
 	scopes               []func(*DB) *DB
+	queryTypes           []bool
```
Comment (a631807682) [inline statement.go:49]: Can we store it in `Settings` ? I haven't tested it, but could something wrong with this ?
```go
tx := db.Scope(...)
go func() {
	tx.Scan(...)
}()
go func() {
	tx.Scan(...)
}()
```

Comment (black-06) [reply statement.go:49]: It has the same problem unless we use thread-safe slice/array

Comment (a631807682) [reply statement.go:49]: > It has the same problem unless we use thread-safe slice/array

Or can we use lock?

**Review by a631807682 — APPROVED** [review]
This is ok for me, but I think since we implemented a list, it would be a bit wasteful if we can only deal with bools. Of course, I don't think of other scenarios that need to be used at the moment, so I think it's ok.

**Comment by jinzhu** [comment]
If this PR is only used to fix the panic issue in the nested scan, I think don't panic but return an error should be enough.

We don't expect users to use it like this, for more details: https://gorm.io/docs/method_chaining.html

Going this close this PR for now, feel free to create a new one if those changes are required in other scenarios.

Thank you.

**Pull request closed without merging**